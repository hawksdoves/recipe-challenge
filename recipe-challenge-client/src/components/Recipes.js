import React from 'react';

import RecipeSearch from './RecipeSearch';
import List from './List';

class Recipes extends React.Component {

    searchRecipes = (event) => {
        const updatedResults = this.props.recipes.filter(recipe => {
            return recipe.name.toLowerCase().search(
                event.target.value.toLowerCase()) !== -1;
            });
        this.setState({ recipeResults: updatedResults});
    }

    state = {
        recipeResults: null
    }

    render() {
        const { recipes, ...rest } = this.props

        return  recipes && recipes.length > 0 ? (
            <div className='recipesContainer'>
                <RecipeSearch searchRecipes={this.searchRecipes} />
                <List recipes={this.state.recipeResults || recipes} {...rest} />
            </div>
        ) :
            <div>Sorry, we currently have no recipes for you</div>
    }
}

export default Recipes;
