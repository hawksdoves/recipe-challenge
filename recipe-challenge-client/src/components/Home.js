import React from 'react';
import { Route, Switch } from 'react-router-dom';
import { Loader } from 'semantic-ui-react';

import DetailedRecipe from './DetailedRecipe';
import Recipes from './Recipes';
import RouteNotFound from './RouteNotFound';

class Home extends React.Component {

    componentDidMount() {
        this.props.getRecipes();
    }

    render() {
        const { recipes, isFetchingRecipes, ...rest } = this.props;

        return !isFetchingRecipes ? (
            <Switch>
                <Route exact path="/" render={(props) => (<Recipes {...rest} recipes={recipes} />)} />
                <Route path="/recipe/:id" render={(props) => (<DetailedRecipe {...rest}  />)} />
                <Route path='*' component={RouteNotFound} />
            </Switch>
        ) :
            <Loader active inline='centered' />
    }
};

export default Home;
