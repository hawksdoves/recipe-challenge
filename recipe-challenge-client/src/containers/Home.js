import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import Home from '../components/Home';

import { addFavourite, removeFavourite, getRecipes } from '../actions';


const mapStateToProps = (state) => {

    return {
        recipes: state.recipes,
        favourites: state.favourites,
        isFetchingRecipes: state.isFetchingRecipes
    };
}

const mapDispatchToProps = (dispatch) => bindActionCreators({ 
    addFavourite, 
    removeFavourite, 
    getRecipes
 }, dispatch)

const HomeContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Home)

export default HomeContainer;
