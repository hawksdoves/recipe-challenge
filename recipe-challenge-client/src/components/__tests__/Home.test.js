import React from 'react';
import { shallow, mount } from 'enzyme';
import { StaticRouter, Route } from 'react-router-dom';

import Recipes from '../Recipes';

const mockResponse = {
    data: {
        "recipes": [
            {
                "id": "1",
                "name": "Lemon Chicken",
                "time": "30",
                "ingredients": [
                    { "name": "Chicken" } ,
                    { "name": "Lemon" }
                ]
            }, 
            {
                "id": "2",
                "name": "Beef Stroganoff",
                "time": "30",
                "ingredients": [
                    { "name": "Beef" },
                    { "name": "Mustard" },
                    { "name": "Mushrooms" }
                ]
            }
        ]
    }
};

jest.mock('axios', () => {
    const mockGetRequest = jest.fn(() => Promise.resolve(mockResponse));

    return {
        get: mockGetRequest
    };
});

import Home from '../Home';

describe('Home', () => {
    
    it('should render all the recipes correctly', () => {
        const wrapper = shallow(<Home recipes={mockResponse.data.recipes} />);
        wrapper.setState({ recipes: mockResponse.data.recipes });

        expect(wrapper).toMatchSnapshot();
    })

    it('should display loader when fetching recipes', () =>{
        const wrapper = shallow(<Home isFetchingRecipes={true} />);

        expect(wrapper).toMatchSnapshot();
    })

    it('calls componentDidMount() lifecycle method which calls getRecipes', () => {
        const mockGetRecipes = jest.fn();
        const componentDidMountSpy = jest.spyOn(Home.prototype, 'componentDidMount');
        const wrapper = mount(
            <StaticRouter location="/" context={{}}>
                <Home recipes={mockResponse.data.recipes} getRecipes={mockGetRecipes} favourites={[]} />
            </StaticRouter>
        );

        expect(componentDidMountSpy).toHaveBeenCalled();
        expect(wrapper.find(Home).props().getRecipes).toHaveBeenCalled();
        expect(wrapper.find(Route).find(Recipes).length).toEqual(1);
    })

});
