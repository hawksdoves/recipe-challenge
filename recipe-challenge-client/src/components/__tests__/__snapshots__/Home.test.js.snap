// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should display loader when fetching recipes 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Loader
    active={true}
    inline="centered"
/>,
  "nodes": Array [
    <Loader
      active={true}
      inline="centered"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Home
        isFetchingRecipes={true}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Home {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "isFetchingRecipes": true,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Loader
          active={true}
          inline="centered"
/>,
        "_debugID": 4,
        "_renderedOutput": <Loader
          active={true}
          inline="centered"
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Home
    isFetchingRecipes={true}
/>,
}
`;

exports[`Home should render all the recipes correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Switch>
    <Route
        exact={true}
        path="/"
        render={[Function]}
    />
    <Route
        path="/recipe/:id"
        render={[Function]}
    />
    <Route
        component={[Function]}
        path="*"
    />
</Switch>,
  "nodes": Array [
    <Switch>
      <Route
            exact={true}
            path="/"
            render={[Function]}
      />
      <Route
            path="/recipe/:id"
            render={[Function]}
      />
      <Route
            component={[Function]}
            path="*"
      />
</Switch>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Home
        recipes={
                Array [
                        Object {
                          "id": "1",
                          "ingredients": Array [
                            Object {
                              "name": "Chicken",
                            },
                            Object {
                              "name": "Lemon",
                            },
                          ],
                          "name": "Lemon Chicken",
                          "time": "30",
                        },
                        Object {
                          "id": "2",
                          "ingredients": Array [
                            Object {
                              "name": "Beef",
                            },
                            Object {
                              "name": "Mustard",
                            },
                            Object {
                              "name": "Mushrooms",
                            },
                          ],
                          "name": "Beef Stroganoff",
                          "time": "30",
                        },
                      ]
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Home {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "recipes": Array [
            Object {
              "id": "1",
              "ingredients": Array [
                Object {
                  "name": "Chicken",
                },
                Object {
                  "name": "Lemon",
                },
              ],
              "name": "Lemon Chicken",
              "time": "30",
            },
            Object {
              "id": "2",
              "ingredients": Array [
                Object {
                  "name": "Beef",
                },
                Object {
                  "name": "Mustard",
                },
                Object {
                  "name": "Mushrooms",
                },
              ],
              "name": "Beef Stroganoff",
              "time": "30",
            },
          ],
        },
        "refs": Object {},
        "state": Object {
          "recipes": Array [
            Object {
              "id": "1",
              "ingredients": Array [
                Object {
                  "name": "Chicken",
                },
                Object {
                  "name": "Lemon",
                },
              ],
              "name": "Lemon Chicken",
              "time": "30",
            },
            Object {
              "id": "2",
              "ingredients": Array [
                Object {
                  "name": "Beef",
                },
                Object {
                  "name": "Mustard",
                },
                Object {
                  "name": "Mushrooms",
                },
              ],
              "name": "Beef Stroganoff",
              "time": "30",
            },
          ],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Switch>
          <Route
                    exact={true}
                    path="/"
                    render={[Function]}
          />
          <Route
                    path="/recipe/:id"
                    render={[Function]}
          />
          <Route
                    component={[Function]}
                    path="*"
          />
</Switch>,
        "_debugID": 2,
        "_renderedOutput": <Switch>
          <Route
                    exact={true}
                    path="/"
                    render={[Function]}
          />
          <Route
                    path="/recipe/:id"
                    render={[Function]}
          />
          <Route
                    component={[Function]}
                    path="*"
          />
</Switch>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Home
    recipes={
        Array [
            Object {
              "id": "1",
              "ingredients": Array [
                Object {
                  "name": "Chicken",
                },
                Object {
                  "name": "Lemon",
                },
              ],
              "name": "Lemon Chicken",
              "time": "30",
            },
            Object {
              "id": "2",
              "ingredients": Array [
                Object {
                  "name": "Beef",
                },
                Object {
                  "name": "Mustard",
                },
                Object {
                  "name": "Mushrooms",
                },
              ],
              "name": "Beef Stroganoff",
              "time": "30",
            },
          ]
    }
/>,
}
`;
